src/sentry/utils/circuit_breaker.py

```diff
@@ -42,3 +78,100 @@ def circuit_breaker_activated(
+
+    Note: The callback MUST NOT handle and then silently swallow exceptions, or else they won't
+    count towards the ciruit-breaking. In other words, this function should be used - along with an
+    `except CircuitBreakerTripped` block - inside the try-except wrapping the callback call:
+
+        try:
+            with_circuit_breaker("fire", play_with_fire, config)
+            # or, if the callback takes arguments:
+            # with_circuit_breaker("fire", lambda: play_with_fire(fuel_type="wood"), config)
+        except CircuitBreakerTripped:
+            logger.log("Once burned, twice shy. No playing with fire for you today. Try again tomorrow.")
+        except BurnException:
+            logger.log("Ouch!")
+
+    The threshold for tripping the circuit breaker and whether to allow bypass requests (and if so,
+    how many) can be set in the `config` argument. See the `CircuitBreakerConfig` class and
+    `CIRCUIT_BREAKER_DEFAULTS`.
+    """
+    config: CircuitBreakerConfig = {**CIRCUIT_BREAKER_DEFAULTS, **(custom_config or {})}
+    error_count_key = ERROR_COUNT_CACHE_KEY(key)
+
+    if _should_call_callback(key, error_count_key, config):
+        return _call_callback(error_count_key, config["error_limit_window"], callback)
+    else:
+        raise CircuitBreakerTripped
+
+
+def _should_call_callback(
+    key: str,
+    error_count_key: str,
+    config: CircuitBreakerConfig,
+) -> bool:
+    error_count = _get_or_set_error_count(error_count_key, config["error_limit_window"])
+    if error_count < config["error_limit"]:
+        return True
+
+    # Limit has been exceeded, check if we should allow any requests to pass through
+    if config["allow_passthrough"]:
+        should_bypass = not ratelimiter.backend.is_limited(
+            PASSTHROUGH_RATELIMIT_KEY(key),
+            limit=config["passthrough_attempts_per_interval"],
+            window=config["passthrough_interval"],
+        )
+        if should_bypass:
+            metrics.incr(f"circuit_breaker.{key}.bypassed")
+            return True
+
+    metrics.incr(f"circuit_breaker.{key}.throttled")
+    return False
+
+
+def _call_callback(error_count_key: str, error_limit_window: int, callback: Callable[P, T]) -> T:
+    try:
+        result = callback()
+    except Exception:
+        _update_error_count(error_count_key, error_limit_window)
+        raise
+    else:
+        _update_error_count(error_count_key, error_limit_window, reset=True)
+        return result
+
+
+def _update_error_count(
+    error_count_key: str,
+    error_limit_window: int,
+    reset: bool = False,
+) -> None:
+    """
+    Increment the count at the given key, unless `reset` is True, in which case, reset the count to 0.
+    """
+    current_count = _get_or_set_error_count(error_count_key, error_limit_window)
+    new_count = 0 if reset else current_count + 1
+    cache.set(error_count_key, new_count, error_limit_window)
```

Comment:
_technically_ this does an extra cache.get_or_set which can be skipped if `reset=True`

this isn't as clean but 🤷‍♀️ 
```suggestion
    new_count = 0 
    if not reset:
      current_count = _get_or_set_error_count(error_count_key, error_limit_window)
      new_count = current_count + 1
    cache.set(error_count_key, new_count, error_limit_window)
```

Reply:
Yeah, that actually just occurred to me as I was looking it over again. It's a tiny bit less readable, but I suppose I could just do `new_count = 0 if reset else _get_or_set_error_count(error_count_key, error_limit_window) + 1`.

Reply:
Or I could split up the ternary:

```python
if reset:
    new_count = 0
else:
    new_count = _get_or_set_error_count(error_count_key, error_limit_window) + 1
```

Verbose but very readable. I'll go with that.